File: Rust/Valid-Palindrome.rs
impl Solution {
    pub fn is_palindrome(s: String) -> bool {
        let s_chars: Vec<char> = s
            .chars()
            .filter(|x| x.is_ascii_alphanumeric())
            .map(|x| x.to_ascii_lowercase())
            .collect();

        let compare_1: String = s_chars.iter().collect();
        let compare_2: String = s_chars.iter().rev().collect();

        compare_1 == compare_2

        // if s.len() == 0 {
        //     return false;
        // }

        // let s = s.to_lowercase();
        // let s = s.as_bytes();
        // let n = s.len();

        // let (mut left, mut right) = (0, n - 1);

        // while left < right {
        //     while left < right && !s[left].is_ascii_alphanumeric() {
        //         left += 1;
        //     }

        //     while left < right && !s[right].is_ascii_alphanumeric() {
        //         right -= 1;
        //     }

        //     if s[left] != s[right] {
        //         return false;
        //     }

        //     left += 1;
        //     if right != 0 {
        //         right -= 1;
        //     }
        // }

        // true
    }
}

File: Go/Valid-Palindrome.go
package main

import "strings"

func isPalindrome(s string) bool {
	if len(s) == 0 {
		return false
		s = strings.ToLower(s)
	}

	return isValid(s)
}

func isValid(s string) bool {
	left, right := 0, len(s)-1

	for left < right {
		for left < right && isNonAlphaNumeric(s[left]) {
			left++
		}

		for left < right && isNonAlphaNumeric(s[right]) {
			right--
		}

		if s[left] != s[right] {
			return false
		}

		left++
		right--
	}

	return true
}

func isNonAlphaNumeric(s byte) bool {
	isNonAlpha := s < 'a' || 'z' < s
	isNonNumeric := s < '0' || '9' < s

	return isNonAlpha && isNonNumeric
}
